Add metrics runtime for Wasmer and debug for Wasmtime

JanFalkin opened the pull request:
First this pull is intentional; sorry about the last false alarm.  I am not certain if this is the best way to introduce variants to the engine.  This pull utilizes some new functionality on wasmer that exposes usage metrics from the wasm engine https://github.com/wasmerio/wasmer-go/commit/586bbfef6447d17456d3160e0b2694cdcf7cde75.  Unfortunately this requires the use of unsafe as we must provide a C callback for the metrics calculation.  In addition to the work for wasmer's metrics, I have also included the code for wasmtime to enable a debug session in the presence of a DWARF enable debugger. Both of these features are included in 2 new "new" functions.  Another possibility would be to provide  construct data such that "new" itself would have enough context to determine how the engine should be created.  The pull will hopefully inspire some discussion and I can adapt the code anyway we decide.

codefromthecrypt commented on engine.go:
```diff
@@ -15,6 +16,8 @@ type (
 	Engine interface {
 		// Name of the engine. Ex. "wazero"
 		Name() string
+		NewWithMetering(code []byte, hostCallHandler HostCallHandler, maxInstructions uint64, pfn unsafe.Pointer) (Module, error)
```
How about putting this in ModuleConfig instead as it can prevent api clash. Also, this sounds like configuration :D

JanFalkin replied:
Agreed. Working on it.

JanFalkin replied:
I am now using a completely different approach, but it's gone

codefromthecrypt commented:
Interesting.

There's an alternative to consider which is to allow wasmer and wasmtime to be configured independently, similar to wazero.

In #34 via @Arsen6331, we added `wazero.EngineWithRuntime(newRuntime NewRuntime)` which allowed runtime-specific functionality without bleeding things that aren't necessarily portable or portable by accident. This also allows you to test in the runtime itself which is probably needed as they can make some incompatible changes sometimes.

This also allows you to push down new functionality before abstracting it, which is handy as abstraction is hard to get right.

codefromthecrypt commented on engine_test_impl.go:
```diff
@@ -0,0 +1,23 @@
+package wapc
+
+// __attribute__((weak))
+// extern unsigned long long meteringFn(unsigned int op);
```
ideally running the main tree tests can be done without CGO also. While we aren't doing it right now, one thing this allows is proving other archs work by building a test binary and running on scratch containers (wazero does this for riscv for example)

JanFalkin replied:
file removed

codefromthecrypt commented:
@JanFalkin ps if you are game with doing the same thing as we did in wazero (make a library-specific hole to configure the other runtimes), I'll offer the labor to make a PR that does that including testing the things that are testable. Lemme know!

pkedy reviewed (commented):
Are these metrics engine/wasmer specific?

If so, what I think I'd like to see is all engine specific options moved to the `Engine` function and use [variadic options](https://petomalina.medium.com/dealing-with-optional-parameters-in-go-9780f9bfbd1d) to customize.

So in the case of wasmer metrics, something like...

```go
var engineInstance = engine{
	// things to set with options
	useMetrics bool
}

type EngineOption func(e * engineInstance)

func Engine(opts ...EngineOption) wapc.Engine {
	e := engineInstance{}
	for _, opt := range opts {
		opt(&e)
	}
	return &e
}

func WithMetrics() EngineOption {
   return func(e * engineInstance) {
      e.useMetrics = true
   }
}
```

Using this pattern, `wazero.EngineWithRuntime(runtime)` could be replaced with `wazero.Engine(wazero.WithRuntime(runtime))` and allows for additional customization in the future.

Thoughts @JanFalkin @codefromthecrypt ?

JanFalkin commented:
> Are these metrics engine/wasmer specific?
> 
> If so, what I think I'd like to see is all engine specific options moved to the `Engine` function and use [variadic options](https://petomalina.medium.com/dealing-with-optional-parameters-in-go-9780f9bfbd1d) to customize.
> 
> So in the case of wasmer metrics, something like...
> 
> ```go
> var engineInstance = engine{
> 	// things to set with options
> 	useMetrics bool
> }
> 
> type EngineOption func(e * engineInstance)
> 
> func Engine(opts ...EngineOption) wapc.Engine {
> 	e := engineInstance{}
> 	for _, opt := range opts {
> 		opt(&e)
> 	}
> 	return &e
> }
> 
> func WithMetrics() EngineOption {
>    return func(e * engineInstance) {
>       e.useMetrics = true
>    }
> }
> ```
> 
> Using this pattern, `wazero.EngineWithRuntime(runtime)` could be replaced with `wazero.Engine(wazero.WithRuntime(runtime))` and allows for additional customization in the future.
> 
> Thoughts @JanFalkin @codefromthecrypt ?

They are wasmer specific and I think I like your idea @pkedy . I will work on that idea today given time.

JanFalkin commented:
@pkedy Does the current look like what you envisioned?

codefromthecrypt commented on engine.go:
```diff
@@ -24,6 +24,11 @@ type (
 		New(ctx context.Context, host HostCallHandler, guest []byte, config *ModuleConfig) (Module, error)
 	}
 
+	MeteringInfo struct {
```
Since this is not an abstract, and possibly unnatural way, to provide metering info in go, what if instead the caller casts the concrete type when they need this? In other words, define this on the only engine that supports it. Then if metering becomes common and in a more natural way (vs unsafe pointers), there's still the ability to abstract it. However, abstracting with only one impl seems not the right time. SG?

codefromthecrypt commented on engines/wazero/wazero.go:
```diff
@@ -321,6 +320,10 @@ func (m *Module) Instantiate(ctx context.Context) (wapc.Instance, error) {
 	return &instance, nil
 }
 
+func (i *Instance) RemainingPoints(context.Context) uint64 {
```
I would remove this from the top-level interface as then you don't need to document no-op on something only one engine supports.

JanFalkin replied:
@codefromthecrypt Let me clean this up a bit more.  I see your points and will try to make it so.

codefromthecrypt reviewed (commented):
I think this is getting closer, but I think the function for wasmer metering should only be defined and implemented on the wasmer interface, then some sort of test added. I think instruction count might become common, but if it did, probably wouldn't look like this, at least not in go. Doing it in the wasmer engine also allows you to document on that func exactly what it is going to be doing.

wdyt?

codefromthecrypt commented:
last comment is to re-title this to "Allow engine specific configuration for wasmer and wasmtime" as this is basically the infrastructure later changes can build on, also. While metering and debug are quite interesting, there are also other features folks may want to configure. It may become more sustainable to make an option for the runtime's "config" type instead of punching a hole for each flag. That said, what's done here is fine for now.

In the case of wazero, we already have the ability to configure the runtime, and through context parameters anything experimental as well. That said, in the future wazero engine might need something similar to this pattern, though we can wait till then.

codefromthecrypt commented on go.mod:
```diff
@@ -5,8 +5,8 @@ go 1.18
 require (
 	github.com/Workiva/go-datastructures v1.0.53
 	github.com/bytecodealliance/wasmtime-go v0.39.0
+	github.com/wasmerio/wasmer-go v1.0.5-0.20220411075136-586bbfef6447
```
ps this is a pre-release. it might be worth nagging wasmer to cut a release as they haven't in over a year, but probably would.

JanFalkin replied:
@codefromthecrypt OK.  I'll give it a try to prod wasmer-go.  My pull was the last to land there and it has been a while.

codefromthecrypt reviewed (commented):
I'd like to help get this landed, so I spent some time looking at the impl in the other repos.

I think how wasmer implements metering is interesting and people should be able to use it. The TL;DR; is I think the underlying wasmer instance type might be the easiest way out, but I think we should keep wasmer vocab and also not corner the instance to only partial functionality.

First and most importantly, we have to concede that even where implemented wasmer and wasmtime at least have different vocab for metering. Ex wasmer it is [points](https://github.com/wasmerio/wasmer/blob/3b67f8d17fbe8dd9dac40b18b66a22e49ee1fe1e/lib/middlewares/src/metering.rs#L401-L405) and wasmtime it is [fuel](https://github.com/bytecodealliance/wasmtime/blob/89ccc56e463bf6c6a1d8cf3f0c2e245861e85c2f/tests/all/fuel.rs). I don't think we can assume these are mapped equally, and if we did we'd want to test that is the case. Meanwhile, we don't have to as we can instead keep the types in the wasmer impl.

Mainly, the unit of configuration wouldn't be "instructions". In other words, max points makes sense for a wasmer-specific impl because they document and test what that is. This allows you to test it here also to ensure folks are configuring properly. It also maps in a non-confusing way to the functions that can modify that or get remaining values at runtime.

On impl, I poked around a bit and there's a mild change I would suggest on the `get_remaining_points` func as it has two items, remaining and exhausted https://github.com/wasmerio/wasmer/blob/3b67f8d17fbe8dd9dac40b18b66a22e49ee1fe1e/lib/middlewares/src/metering.rs#L113-L123

In go, we should be able to map that to multiple results,`(uint64, bool)` to cover both.

Finally, there are other functions about metering in wasmer. ex set_remaining_points. For this reason I would suggest having `wasmer.Instance` here return `.Metering()` which when nil (or, false) means disabled but otherwise starts with the remaining points function, but can have the other functions added later.

That said, a much simpler way could be to return wasmer.Instance.Unwrap() or .Raw() and just return the underlying wasmer instance type. This prevents you from having to punch a hole feature by feature which can affect compatibility and also lock up people from being able to test something before updating wapc first.

hope this helps! and cc @syrusakbary for review!

codefromthecrypt commented:
one other thing, and this is mostly my curiosity, but it might be handy when documenting this to give the use case. For example, I would guess this is to allow host functions to early exit? If not, would this would be a different goroutine watching values? If so, maybe mention any value visibility concerns if there are any (ex if multiple goroutines, they could be on different OS threads and possibly not read accurate values). This could also be done in the wasmer-go lib and simply referred back to there, or dismissed as too niche to matter ;)

codefromthecrypt commented on engines/wasmer/wasmer.go:
```diff
@@ -95,10 +100,22 @@ type (
+	return &e
+}
+
+func WithMetrics(maxInstructions uint64, pfn unsafe.Pointer) EngineOption {
```
We can probably be more specific, namely this is mostly about metering, right?
https://github.com/wasmerio/wasmer/blob/ea12ec0a0646325500588b42c48ac91a252649ae/lib/c-api/src/wasm_c_api/unstable/middlewares/metering.rs#L176-L179

```suggestion
func WithPushMeteringMiddlewarePtr(initialLimit uint64, pfn unsafe.Pointer) EngineOption {
```

PS this is one reason why I think it might be better to instead expose `WithConfig` as then you don't need to worry about the syntax and whether any typos matching config vals. Plus signature changes later don't end up rev-locking as most folks config types themselves don't move even if the values in them do.

JanFalkin replied:
I would consider WithConfig better than WithPushMeteringMiddlewarePtr as the later is more an implementation detail. As for the limit, I am not certain how termination occurs and will try to find out exactly.  My main goal for this is to enable the metering which will allow our fabric to bill tenants on usage.  Termination for us will occur is there is insufficient blockchain gas regardless. So the termination is not my concern but I will try to find it for others sake.

codefromthecrypt replied:
thanks for the clarification and the use case description! (ps I think what mainly got me into rename mode was metrics -> metering)

JanFalkin commented:
> I'd like to help get this landed, so I spent some time looking at the impl in the other repos.
> 
> I think how wasmer implements metering is interesting and people should be able to use it. The TL;DR; is I think the underlying wasmer instance type might be the easiest way out, but I think we should keep wasmer vocab and also not corner the instance to only partial functionality.
> 
> First and most importantly, we have to concede that even where implemented wasmer and wasmtime at least have different vocab for metering. Ex wasmer it is [points](https://github.com/wasmerio/wasmer/blob/3b67f8d17fbe8dd9dac40b18b66a22e49ee1fe1e/lib/middlewares/src/metering.rs#L401-L405) and wasmtime it is [fuel](https://github.com/bytecodealliance/wasmtime/blob/89ccc56e463bf6c6a1d8cf3f0c2e245861e85c2f/tests/all/fuel.rs). I don't think we can assume these are mapped equally, and if we did we'd want to test that is the case. Meanwhile, we don't have to as we can instead keep the types in the wasmer impl.
> 
> Mainly, the unit of configuration wouldn't be "instructions". In other words, max points makes sense for a wasmer-specific impl because they document and test what that is. This allows you to test it here also to ensure folks are configuring properly. It also maps in a non-confusing way to the functions that can modify that or get remaining values at runtime.
> 
> On impl, I poked around a bit and there's a mild change I would suggest on the `get_remaining_points` func as it has two items, remaining and exhausted https://github.com/wasmerio/wasmer/blob/3b67f8d17fbe8dd9dac40b18b66a22e49ee1fe1e/lib/middlewares/src/metering.rs#L113-L123
> 
> In go, we should be able to map that to multiple results,`(uint64, bool)` to cover both.
> 
> Finally, there are other functions about metering in wasmer. ex set_remaining_points. For this reason I would suggest having `wasmer.Instance` here return `.Metering()` which when nil (or, false) means disabled but otherwise starts with the remaining points function, but can have the other functions added later.
> 
> That said, a much simpler way could be to return wasmer.Instance.Unwrap() or .Raw() and just return the underlying wasmer instance type. This prevents you from having to punch a hole feature by feature which can affect compatibility and also lock up people from being able to test something before updating wapc first.
> 
> hope this helps! and cc @syrusakbary for review!

I am looking at adding the bool to RemainingPoints such that we get (uint64,bool).  The problem I see is that wasmtime's meaning of the bool is much different than wasmer's.  wasmtime treats that bool as a check for whether the store has a fuel limit set while wasmer's is to say that fuel remaining is 0 and execution has been terminated.  So I can create the new signature and have it return values but they will have a different meaning across the engines.  Is that desirable @codefromthecrypt ?

codefromthecrypt commented on engines/wasmtime/wasmtime.go:
```diff
@@ -75,10 +79,26 @@ type (
+	}
+}
+
+func WithMetrics(maxGas uint64) EngineOption {
```
isn't the word for something like this metering, not metrics? metrics usually is different, like perf metrics or error count.

JanFalkin replied:
I'm changing these to WithMetering.

codefromthecrypt commented on engines/wasmer/wasmer.go:
```diff
@@ -611,3 +636,12 @@ func (m *Module) Close(context.Context) error {
 	return nil
 }
+
+func (i *Instance) RemainingPoints(context.Context) (uint64, bool) {
```
```suggestion
// RemainingPoints returns the amount of points remaining or false if exhausted.
func (i *Instance) RemainingPoints(context.Context) (uint64, bool) {
```

codefromthecrypt commented on engines/wasmtime/wasmtime.go:
```diff
@@ -87,9 +107,21 @@ func (e *engine) Name() string {
 	store := wasmtime.NewStore(engine)
-
+	if e.useMetrics {
+		err := store.AddFuel(e.maxGas)
```
maybe settle on one noun, seems wasmer uses both fuel and points. I think we shouldn't add a third word, gas? Also, if you pass the config instead then you don't have to decide what to name it.

JanFalkin replied:
Agreed.  My bad.  Started with trying to normalize all to metrics only to reverse course.  I will make thing internally consistent.

JanFalkin replied:
wasmtime actually uses the term gas in their docs for their metering support.  That was where I got it from.  I was thinking that if we use the native terms here it will indicate to the caller that there are subtle differences that should not be ignored.

codefromthecrypt replied:
maybe put this in a comment to prevent confusion?
```suggestion
		// wasmtime uses the word gas in docs even though their API calls it fuel.
		err := store.AddFuel(e.maxGas)
```

JanFalkin replied:
My bad.  We use the term gas in the blockchain world which is where gas came from.  Changing to fuel.  Sorry for the confusion

codefromthecrypt commented on engines/wasmtime/wasmtime.go:
```diff
@@ -108,6 +140,10 @@ func (e *engine) New(_ context.Context, host wapc.HostCallHandler, guest []byte,
 	}, nil
 }
 
+func (i *Instance) FuelConsumed(context.Context) (uint64, bool) {
```
```suggestion
// FuelConsumed returns the amount of fuel consumed in the underlying store or false if fuel consumption was not configured.
func (i *Instance) FuelConsumed(context.Context) (uint64, bool) {
```

codefromthecrypt reviewed (commented):
I think it is ok for these to be named different as they are engine-specific and you literally can't have both running at the same time due to wasi conflicts (remember these are driven by build tags, not config tags).


Main thing is to add an Example test for each which shows how to use this (also godoc which mainly just summarizes why this is used).

Ex. and forgive any typos
```go
// This shows how to read the fuel consumed in the current wasmer store.
func Example_fuelConsumed() {
	ctx := context.Background()

	// setup the instance

	if w, isWasmtime := instance.(*wasmtime.Instance); !isWasmtime {
		// maybe wazero or wasmer
	} else if fuel, enabled := w.FuelConsumed(ctx); !enabled {
		// Config.SetConsumeFuel wasn't configured	 
	} else if fuel > 0 {
		// Don't print real value as it may vary and break the test!
		log.Println("fuel used")
	} else {
		log.Println("no fuel used, yet")
	}

	// Output:
	// fuel used
}
```

codefromthecrypt commented on engines/wasmer/wasmer.go:
```diff
@@ -95,10 +100,22 @@ type (
+	return &e
+}
+
+func WithMetering(maxInstructions uint64, pfn unsafe.Pointer) EngineOption {
```
if this change continues to poke holes for specific config instead of passing wasmer.Config, then we'll need to have docs on these functions. ex refer to PushMeteringMiddlewarePtr as otherwise it is hard for someone who doesn't know yet to understand what pfn means. Regardless, best to add an example test as show>tell

JanFalkin replied:
Example is next, I promise!

JanFalkin replied:
I hope the new godoc above was what was envisioned

codefromthecrypt commented:
thanks. I will raise a branch off yours to show what I mean with example. you almost got it!

codefromthecrypt commented:
here's what I was getting at: to ensure the the code in the example works, you can use `Example` tests which godoc will automatically embed.

https://github.com/JanFalkin/wapc-go/compare/master...codefromthecrypt:wapc-go:example_test

but as you will notice trying to use CGO in tests is quite gnarly. I feel the maintenance of this current approach is hinted at due to the amount of comments. I would like to raise an alternate PR with the approach I mentioned earlier, as it is IMHO far more viable, as well less complicated and coupled to expose the underlying module vs punch a hole per config. I hope you'll have a look and understand the main goal is to get you working, and this is more important than a specific way to do that. If we can get you working with less support load, that's all the better as we can divert the load otherwise spent on drift management between library and abstraction in other ways.

codefromthecrypt commented:
@JanFalkin PTAL as sometimes show > tell and I think if you try this, you'll see what I mean. Notably, you don't need to change waPC to a pre-release version of wasmer in order to use features added after the one here. All you need to do is update your local go.mod https://github.com/wapc/wapc-go/pull/41

If works for you, I'll do the other engines similarly so that things are consistent cc also @pkedy

JanFalkin commented:
> @JanFalkin PTAL as sometimes show > tell and I think if you try this, you'll see what I mean. Notably, you don't need to change waPC to a pre-release version of wasmer in order to use features added after the one here. All you need to do is update your local go.mod #41
> 
> If works for you, I'll do the other engines similarly so that things are consistent cc also @pkedy
@codefromthecrypt I pulled in your branch and also modified wasmtime.  Look about right now?

JanFalkin commented:
@pkedy @codefromthecrypt after careful thought I removed the interface parameter not necessary. I also restored 1.0.4 for wasmer so no change to go.mod.  There is some duplication in the tests.  Given a few more cycles I will try to remove.  I also added wasmtime as appropriate.

codefromthecrypt commented on engines/wasmtime/wasmtime.go:
```diff
@@ -108,6 +124,11 @@ func (e *engine) New(_ context.Context, host wapc.HostCallHandler, guest []byte,
 }
 
+// FuelConsumed returns the amount of fuel consumed in the underlying store or false if fuel consumption was not configured.
+func (i *Instance) FuelConsumed(context.Context) (uint64, bool) {
```
how about moving this to Module.UnwrapStore() instead? then you don't need to do things like this for other features

JanFalkin replied:
I am adding an UnwrapStore under Module and Instance.  It is more useful to me under Instance. Also added to wasmer as well as long as I was here anyway.

codefromthecrypt replied:
sounds fine

codefromthecrypt reviewed (commented):
lookin good

pkedy commented on engines/wasmtime/wasmtime.go:
```diff
@@ -75,10 +77,22 @@ type (
+
+// WithEngine allows you to override the default engine. Defaults to
+// `wasmer.NewEngine`.
+func WithEngine(engineFn func() *wasmtime.Engine) EngineOption {
```
Just curious: Why pass in a function to create the engine instead of passing in an engine instance?

JanFalkin replied:
@pkedy Actually I don't see any reason other than an example I worked from may have used a creator func rather than the engine itself.  Is it better with passing the engine or just different as I think I can easily (famous last words) rework.

codefromthecrypt replied:
note if we pass the engine instead, it becomes the same as existing wazero code which passes the runtime

```go
// EngineWithRuntime allows you to customize or return an alternative to
// DefaultRuntime,
func EngineWithRuntime(newRuntime NewRuntime) wapc.Engine {
	return &engine{newRuntime: newRuntime}
}
```

except wazero is missing the unwrap functions added here to the other two..

codefromthecrypt replied:
one thing that may come up later is customization of instantiation of modules within a runtime-specific way beyond the portable `wapc.ModuleConfig`, but in that case one way out is to change it from a struct to an interface, in any case we can solve it when we get to it.

codefromthecrypt reviewed (commented):
Food for thought. In wazero, the engine creation is deferrable, which allows you to decide whether to instantiate a new runtime per module or not. I'm not entirely sure how safe it is on the other runtimes to re-use their engine, especially in parallel.

Ex.
```go
e := EngineWithRuntime(func(ctx context.Context) (wazero.Runtime, error) {
	return r, nil
})
```

JanFalkin commented:
> Food for thought. In wazero, the engine creation is deferrable, which allows you to decide whether to instantiate a new runtime per module or not. I'm not entirely sure how safe it is on the other runtimes to re-use their engine, especially in parallel.
> 
> Ex.
> 
> ```go
> e := EngineWithRuntime(func(ctx context.Context) (wazero.Runtime, error) {
> 	return r, nil
> })
> ```
Based on the way we have it now, the engine instance is passed, so the same model could be tried.  I am not certain at all that wasmer or wasmtime is safe to use in those instances.  That would be for the user to protect by either getting new instances always or somehow safeguarding thread access and synchronize the threads such that only a single passes.

codefromthecrypt commented:
I don't think a user should have to protect calls to the engine. This would be unnecessary for safe ones. Maybe a benchmark can be used, if any reason just to show if something crashes or not? 

Ex. I recall this because wazero tests ended up crashing on both wasmtime and wasmer. Basically I added a benchmark and then moved things to smaller scopes until they stopped crashing. If you do locking instead, then effectively you get no affect from pooling.

https://github.com/tetratelabs/wazero/blob/main/internal/integration_test/vs/wasmtime/wasmtime.go#L47-L48
https://github.com/tetratelabs/wazero/blob/main/internal/integration_test/vs/wasmer/wasmer.go#L56-L57


In any case the main advice is to change the structure but not the scope in this change. Ex change how it is initialized, but don't make it share an engine as the current code doesn't. If you want to make the change to have wasmer and wasmtime share an instance, maybe do that in another commit so it is easier to recognize and back-out (ex do it in a commit after this for example)

codefromthecrypt commented on engines/wasmer/wasmer.go:
```diff
@@ -95,28 +97,43 @@ type (
+	return engine
 }
 
+func Engine(engs ...*wasmer.Engine) wapc.Engine {
```
I think you meant to use the func type

```suggestion
func Engine(options ...EngineOption) wapc.Engine {
```

JanFalkin replied:
I thought I had it as such and there was feedback from either @pkedy or @syrusakbary that had me switch to the specific engine*.  I can put it back; I just want to make sure we don't go back and forth. Thanks

codefromthecrypt replied:
yep. if the preference is to remove the option type I think it still exists in the file, so should also be removed from there.

JanFalkin replied:
I see.  Doing so now.

JanFalkin replied:
Sorry ran out of time today.  Will commit hopefully Sat night or Sunday.

JanFalkin replied:
@codefromthecrypt Sorry about the delay; I was rushing on Friday night.  I think the change was actually very small.  Let me know if this is not what you were asking for.

codefromthecrypt commented on engines/wasmtime/wasmtime.go:
```diff
@@ -75,10 +77,20 @@ type (
 
-func Engine() wapc.Engine {
-	return &engineInstance
+func Engine(engs ...*wasmtime.Engine) wapc.Engine {
```
I think you meant to define a func type:

```suggestion
// EngineOption is a wasmtime-specific configuration, applied via Engine
type Option func(*engine)

func Engine(options ...EngineOption) wapc.Engine {
```

JanFalkin replied:
see https://github.com/wapc/wapc-go/pull/40#discussion_r985022623

codefromthecrypt reviewed (commented):
My feedback remains the same that the option for creating an engine should be a function of it (like in wasmtime) as that ensures the engine returned can close it. that matches the scope of the prior code, but still gives flexibility for the person using it to return a function of a constant engine.

PS whatever syntax here should also be applied to wazero as it would feel off otherwise. I can do that after the fact once things settle.

codefromthecrypt commented on engines/wasmer/wasmer.go:
```diff
@@ -95,28 +97,40 @@ type (
+	return engine
 }
 
+func Engine(engs ...*wasmer.Engine) wapc.Engine {
```
this feels a bit odd, is this common (using varargs to make one of the args optional)?

I think how wazero does it is a lot less confusing (having a second function vs being too clever with one)

```
func EngineWithRuntime(newRuntime NewRuntime) wapc.Engine {
```

JanFalkin replied:
@codefromthecrypt I was looking at that too but, if I remember, there was a comment regarding removing the function and just passing the engine ptr.  Without using the varargs syntax you would have to pass nil if you want a new engine.  I thought that was hideous looking but I am certainly willing to make it a function as orig, I just don't want to reignite the comment of removing the function for the ptr. I can make it look like wazero if that is desirable?

JanFalkin replied:
Works for me!  I'll give it a whirl.

JanFalkin replied:
@codefromthecrypt.  Changes pushed

codefromthecrypt commented on engines/wasmtime/wasmtime.go:
```diff
@@ -464,6 +494,16 @@ func (i *Instance) Close(context.Context) error {
+}
+
+// UnwrapStore allows access to wasmtime-specific features.
+func (m *Module) UnwrapStore() *wasmtime.Store {
```
maybe add functions like this to wazero for parity, noting that the naming suffix should reflect the type returned and wazero doesn't call things store :D

JanFalkin replied:
Makes sense, will do.

codefromthecrypt commented on engines/wasmer/example_test.go:
```diff
@@ -0,0 +1,30 @@
+	ctx := context.Background()
+
+	// Configure waPC to use a specific wasmer feature.
+	e := EngineWithRuntime(func() (*wasmer.Engine, error) {
```
someone may feel like this should be `EngineWithEngine` to match the type, but I feel runtime is nicer than stuttering.

JanFalkin replied:
Agreed.  I was thinking along those lines as well, but the name is cumbersome to say the least.

JanFalkin commented on engines/wazero/wazero.go:
```diff
@@ -145,6 +144,16 @@ func (e *engine) New(ctx context.Context, host wapc.HostCallHandler, guest []byt
 }
 
+// UnwrapCompiledModule allows access to wazero-specific compiled module.
+func (m *Module) UnwrapCompiledModule() *wazero.CompiledModule {
```
done and pushed

pkedy commented:
@JanFalkin I've been a bit busy but rest assured I'll get to this soon.

JanFalkin commented:
> @JanFalkin I've been a bit busy but rest assured I'll get to this soon.

BTW I am away starting this evening till tomorrow evening.  In case you are looking for me.

pkedy commented on engines/wasmer/example_test.go:
```diff
@@ -0,0 +1,30 @@
+package wasmer
```
Please add this build tag. I have an M1 Mac apparently this feature is not supported.

```go
//go:build amd64
// +build amd64
```

pkedy commented on engines/wasmer/wasmer.go:
```diff
@@ -95,28 +97,47 @@ type (
+var engineInstance = engine{newRuntime: DefaultRuntime}
 
+// Engine returns a new wapc.Engine which uses the DefaultRuntime.
 func Engine() wapc.Engine {
```
Instead of having separate `Engine` and `EngineWithRuntime` functions, I wonder about using the Functional Options pattern here. It would be backward compatible. Is there any reason why we wouldn't use it?

https://golang.cafe/blog/golang-functional-options-pattern.html

pkedy replied:
This can be consistent and backward compatible with functional options, but thinking about it more there might be an advantage to having the signature for `Engine` be the same (passing to higher order functions). Let's keep this as is.

pkedy reviewed (commented):
Just 2 small details

pkedy reviewed (approved):
LGTM!

@JanFalkin Thank you for your unwavering engagement on this!
@codefromthecrypt Thank you for all the feedback!

codefromthecrypt commented:
@JanFalkin do you have any feedback since this is working? Are you using the feature successfully and how do you establish the cost and limits you use, if so. curious as we think about this generically.

codefromthecrypt commented:
guess it would be a follow-up to https://github.com/tetratelabs/wazero/issues/422#issuecomment-1140054395 about if things are working now or not, and if so any insights since putting it together.

JanFalkin commented:
@codefromthecrypt Stuff is working as expected.  I have 2 scenarios that require the feature.  The wasmer feature unfortunately has not been tested to a great extent however we have been using the dwarf debugging capabilities of wasmtime a lot.  The metering is being tested at a basic level but we have not fully enabled the wasm engine in our demo and prod environments yet.  I am ahead of the curve a bit.
